File: Bruteforce/codeforces_1995B1.py

# Problem link ===>> https://codeforces.com/problemset/problem/1995/B1

MOD = 10**9 + 7


def pow_mod(x, q, mod):
    if q == 0:
        return 1
    i = pow_mod(x, q // 2, mod)
    if q % 2 == 0:
        return (i * i) % mod
    return (i * i * x) % mod


def solve():
    n, m = map(int, input().split())
    a = list(map(int, input().split()))
    hs = {}
    for x in a:
        hs[x] = hs.get(x, 0) + 1
    ans = 0
    for x in hs:
        y = hs[x]
        l = min(m // x, y)
        ans = max(ans, l * x)
        if x + 1 not in hs:
            continue
        z = hs[x + 1]
        for i in range(1, y + 1):
            if i * x > m:
                break
            du = m - i * x
            su = min(du // (x + 1), z) * (x + 1) + i * x
            ans = max(su, ans)
    print(ans)


tests = int(input())
for _ in range(tests):
    solve()

File: Bruteforce/codeforces_1995B1.cpp

// Problem link ===>> https://codeforces.com/problemset/problem/1995/B1
// submission link ===>>

#include <bits/stdc++.h>
#define ll long long
using namespace std;
const int MOD = 1e9 + 7;

long long pow_mod(long long x, long long q, long long mod)
{
    if (q == 0)
        return 1;
    long long i = pow_mod(x, q / 2, mod);
    if (q % 2 == 0)
        return (i * i) % mod;
    return (i * i * x) % mod;
}

int main()
{
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);

    ll t;
    cin >> t;
    while (t--)
    {
        ll n, m;
        cin >> n >> m;

        vector<int> a(n);
        for (int i = 0; i < n; ++i)
        {
            cin >> a[i];
        }
        map<int, int> hs;
        for (int x : a)
        {
            hs[x]++;
        }
        long long ans = 0;
        for (auto &[x, y] : hs)
        {
            int l = min(m / x, y);
            ans = max(ans, (long long)l * x);
            if (hs.find(x + 1) == hs.end())
                continue;
            int z = hs[x + 1];
            for (int i = 1; i <= y; ++i)
            {
                if (i * x > m)
                    break;
                int du = m - i * x;
                long long su = min(du / (x + 1), z) * (x + 1) + i * x;
                ans = max(ans, su);
            }
        }
        cout << ans << endl;
    }
    return 0;
}
